miniapp/miniapp_triangular_solver.cpp

```diff
@@ -0,0 +1,210 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <iostream>
+
+#include <mpi.h>
+#include <hpx/hpx.hpp>
+
+#include "dlaf/common/index2d.h"
+#include "dlaf/communication/communicator_grid.h"
+#include "dlaf/communication/init.h"
+#include "dlaf/matrix.h"
+#include "dlaf/matrix/index.h"
+#include "dlaf/solver/mc.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+#include "dlaf_test/matrix/util_generic_blas.h"
+#include "dlaf_test/matrix/util_matrix.h"
+#include "dlaf_test/util_types.h"
+
+#include "dlaf/common/timer.h"
+
+namespace {
+
+using dlaf::Backend;
+using dlaf::Device;
+using dlaf::GlobalElementIndex;
+using dlaf::GlobalElementSize;
+using dlaf::SizeType;
+using dlaf::TileElementSize;
+using dlaf::comm::Communicator;
+using dlaf::comm::CommunicatorGrid;
+using dlaf::common::Ordering;
+
+using dlaf_test::TypeUtilities;
+
+using T = double;
+using MatrixType = dlaf::Matrix<T, Device::CPU>;
+
+struct options_t {
+  SizeType m;
+  SizeType mb;
+  SizeType n;
+  SizeType nb;
+  int grid_rows;
+  int grid_cols;
+  int64_t nruns;
+  bool do_check;
+};
+
+options_t check_options(hpx::program_options::variables_map& vm);
+
+void waitall_tiles(MatrixType& matrix);
+
+}
+
+int hpx_main(hpx::program_options::variables_map& vm) {
+  options_t opts = check_options(vm);
+
+  Communicator world(MPI_COMM_WORLD);
+  CommunicatorGrid comm_grid(world, opts.grid_rows, opts.grid_cols, Ordering::ColumnMajor);
+
+  // Allocate memory for the matrices
+  MatrixType A(GlobalElementSize{opts.m, opts.m}, TileElementSize{opts.mb, opts.mb}, comm_grid);
+  MatrixType b(GlobalElementSize{opts.m, opts.n}, TileElementSize{opts.mb, opts.nb}, comm_grid);
+
+  const auto side = blas::Side::Left;
+  const auto uplo = blas::Uplo::Lower;
+  const auto op = blas::Op::NoTrans;
+  const auto diag = blas::Diag::NonUnit;
+  const T alpha = 2.0;
+
+  double m = A.size().rows();
+  double n = b.size().cols();
+  auto add_mul = n * m * m / 2;
+  const double total_ops = dlaf::total_ops<T>(add_mul, add_mul);
+
+  using dlaf::matrix::test::getLeftTriangularSystem;
+  std::function<T(const GlobalElementIndex&)> setter_A, setter_b, expected_b;
+  std::tie(setter_A, setter_b, expected_b) =
+      getLeftTriangularSystem<GlobalElementIndex, T>(uplo, op, diag, alpha, A.size().rows());
```

Comment:
I suspect that setup will hit the subnormal problem as well.

Reply:
Experimentally, checking with `perf` for subnormal interrupts and by looking at single node performances, it does not suffer (with m=10240, n=10240). I'm checking how values of the reference linear systems are generated (I will share with you the details) and we should be on the safe side.